src/Build/Evaluation/ItemSpec.cs

```diff
@@ -6,37 +6,169 @@
-    /// provides some operations over them (like matching items against a given ItemSpec)
+    ///     Represents the elements of an item specification string (e.g. Include="*.cs;foo;@(i)") and
+    ///     provides some operations over them (like matching items against a given ItemSpec)
     /// </summary>
     internal class ItemSpec<P, I>
         where P : class, IProperty
-        where I : class, IItem
+        where I : class, IItem, IMetadataTable
     {
+        internal readonly struct ReferencedItem
+        {
+            public I Item { get; }
+            public ValueFragment ItemAsValueFragment { get; }
+
+            public ReferencedItem(I item, ValueFragment itemAsValueFragment)
+            {
+                Item = item;
+                ItemAsValueFragment = itemAsValueFragment;
+            }
+        }
+
+        internal class ItemExpressionFragment : ItemSpecFragment
+        {
+            private readonly ItemSpec<P, I> _containingItemSpec;
+            private Expander<P, I> _expander;
+
+            private IMSBuildGlob _msbuildGlob;
+
+            private List<ReferencedItem> _referencedItems;
+            public ExpressionShredder.ItemExpressionCapture Capture { get; }
+
+            public List<ReferencedItem> ReferencedItems
+            {
+                get
+                {
+                    InitReferencedItemsIfNecessary();
+                    return _referencedItems;
+                }
+            }
+
+            protected override IMSBuildGlob MsBuildGlob
+            {
+                get
+                {
+                    if (InitReferencedItemsIfNecessary() || _msbuildGlob == null)
+                    {
+                        _msbuildGlob = CreateMsBuildGlob();
+                    }
+
+                    return _msbuildGlob;
+                }
+            }
+
+            public ItemExpressionFragment(
+                ExpressionShredder.ItemExpressionCapture capture,
+                string textFragment,
+                ItemSpec<P, I> containingItemSpec,
+                string projectDirectory)
+                : base(textFragment, projectDirectory)
+            {
+                Capture = capture;
+
+                _containingItemSpec = containingItemSpec;
+                _expander = _containingItemSpec.Expander;
+            }
+
+            public override int MatchCount(string itemToMatch)
+            {
+                return ReferencedItems.Count(v => v.ItemAsValueFragment.MatchCount(itemToMatch) > 0);
+            }
+
+            public override bool IsMatch(string itemToMatch)
+            {
+                return ReferencedItems.Any(v => v.ItemAsValueFragment.IsMatch(itemToMatch));
+            }
+
+            public override bool IsMatchOnMetadata(IItem item, IEnumerable<string> metadata, MatchOnMetadataOptions options)
+            {
+                if (options.Equals(MatchOnMetadataOptions.CaseSensitive))
```

Comment:
My motivation for keeping them as separate cases was clarity- combining them makes for a long confusing string of linq that I was trying to avoid... but if everyone thinks it's clear enough as one case I'll change it.

Reply:
You can use the switch statement, instead of if elses, where the sensitive/insensitive cases are shared code, then path like one is different. Also add a default case that throws `ErrorUtilities.ThrowInternalErrorUnreachable`

Reply:
Thanks for bumping this, I don't know how I missed it twice 😆